homeassistant/components/thethingsnetwork/sensor.py

```diff
@@ -1,165 +1,51 @@
-"""Support for The Things Network's Data storage integration."""
-
-from __future__ import annotations
-
-import asyncio
-from http import HTTPStatus
+"""The Things Network's integration sensors."""
 import logging
 
-import aiohttp
-from aiohttp.hdrs import ACCEPT, AUTHORIZATION
-import voluptuous as vol
+from ttn_client import TTNSensorValue
 
-from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity
-from homeassistant.const import (
-    ATTR_DEVICE_ID,
-    ATTR_TIME,
-    CONF_DEVICE_ID,
-    CONTENT_TYPE_JSON,
-)
+from homeassistant.components.sensor import SensorEntity
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
-from homeassistant.helpers.aiohttp_client import async_get_clientsession
-import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
-from . import DATA_TTN, TTN_ACCESS_KEY, TTN_APP_ID, TTN_DATA_STORAGE_URL
+from .const import DOMAIN, ENTRY_DATA_COORDINATOR
+from .entity import TTNEntity
 
 _LOGGER = logging.getLogger(__name__)
 
-ATTR_RAW = "raw"
-
-DEFAULT_TIMEOUT = 10
-CONF_VALUES = "values"
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_DEVICE_ID): cv.string,
-        vol.Required(CONF_VALUES): {cv.string: cv.string},
-    }
-)
-
-
-async def async_setup_platform(
-    hass: HomeAssistant,
-    config: ConfigType,
-    async_add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType | None = None,
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
-    """Set up The Things Network Data storage sensors."""
-    ttn = hass.data[DATA_TTN]
-    device_id = config[CONF_DEVICE_ID]
-    values = config[CONF_VALUES]
-    app_id = ttn.get(TTN_APP_ID)
-    access_key = ttn.get(TTN_ACCESS_KEY)
-
-    ttn_data_storage = TtnDataStorage(hass, app_id, device_id, access_key, values)
-    success = await ttn_data_storage.async_update()
+    """Add entities for TTN."""
 
-    if not success:
-        return
+    coordinator = hass.data[DOMAIN][entry.entry_id][ENTRY_DATA_COORDINATOR]
 
-    devices = []
-    for value, unit_of_measurement in values.items():
-        devices.append(
-            TtnDataSensor(ttn_data_storage, device_id, value, unit_of_measurement)
-        )
-    async_add_entities(devices, True)
+    sensors: dict[str, TtnDataSensor] = {}
 
+    def _async_measurement_listener() -> None:
+        data = coordinator.data
+        new_sensors = {
+            unique_id: TtnDataSensor(
+                coordinator,
+                ttn_value,
+            )
+            for device_id, device_uplinks in data.items()
+            for field_id, ttn_value in device_uplinks.items()
+            for unique_id in set(TtnDataSensor.get_unique_id(device_id, field_id))
```

Comment:
Why do we do this?

Reply:
see comment in get_unique_id. But I thought I could just store the created entries by the device_id and field_id as we do not need to know the unique_id.